### PR TITLE
Device: Add support for discovering multiple `unix-hotplug` devices

### PR DESCRIPTION
**Pull request by kadinsayani**
Fixes 1/2 related issues raised in https://github.com/canonical/lxd/issues/14266. This fix has been tested locally with a USB hard disk and by @jonathan-conder with USB camera devices.

Summary of changes:
- Adds support for discovering multiple `unix-hotplug` devices which share the same `productid` and/or `vendorid`;
- Improves `productid` and `vendorid` matching logic;
- Resolves inconsistent device matching bug.

### CODE REVIEW

File: lxd/device/unix_hotplug.go [hunk 1]
```diff
@@ -149,29 +149,31 @@ func (d *unixHotplug) Start() (*deviceConfig.RunConfig, error) {
...
+		if device.Subsystem() == "block" {
+			err = unixDeviceSetupBlockNum(d.state, d.inst.DevicesPath(), "unix", d.name, d.config, major, minor, device.Devnode(), false, &runConf)
+		} else {
+			err = unixDeviceSetupCharNum(d.state, d.inst.DevicesPath(), "unix", d.name, d.config, major, minor, device.Devnode(), false, &runConf)
+		}
 
-	if err != nil {
-		return nil, err
+		if err != nil {
+			return nil, err
```
Comment (tomponline) [inline lxd/device/unix_hotplug.go:175]: Do we need to consider undoing any previously setup devices if one of the subsequent device setups fails here?

Comment (tomponline) [reply lxd/device/unix_hotplug.go:175]: I.e. doing the equivalent of when the instance is stopped.

File: lxd/device/unix_hotplug.go [hunk 2]
```diff
@@ -230,6 +232,7 @@ func (d *unixHotplug) loadUnixDevice() *udev.Device {
 	}
 
 	devices, _ := e.Devices()
+	var matchingDevices []*udev.Device
 	var device *udev.Device
```
Comment (tomponline) [inline lxd/device/unix_hotplug.go:236]: I wonder why device var here is outside of the for loop, it should be just:

```go
device := devices[i]
```

Comment (kadinsayani) [reply lxd/device/unix_hotplug.go:236]: Agreed

File: lxd/device/unix_hotplug.go (hunk 1 above)
Comment (tomponline) [reply lxd/device/unix_hotplug.go:175]: But ur doing this in a for loop which means we need to undo files created in earlier iterations if we encounter an error and return

Comment (kadinsayani) [reply lxd/device/unix_hotplug.go:175]: Gotcha, makes sense!

**Comment by kadinsayani** [comment]
I was trying to reason about why the initial implementation of `loadUnixDevice()` only matched the first device with a matching product and vendor id since this logic is explicitly described in the go doc comment. I found the original PR and this comment: https://github.com/canonical/lxd/pull/6530#issuecomment-561348982. 

I don't think the approach in this PR is quite right. Furthermore, wrt https://github.com/canonical/lxd/issues/14373, I found this comment which provides reason to not add support for specifying subsystem types: https://github.com/canonical/lxd/pull/6530#issuecomment-561466424.

We'll have to rethink our solution here.

**Comment by tomponline** [comment]
> I was trying to reason about why the initial implementation of `loadUnixDevice()` only matched the first device with a matching product and vendor id since this logic is explicitly described in the go doc comment. I found the original PR and this comment: [#6530 (comment)](https://github.com/canonical/lxd/pull/6530#issuecomment-561348982).
> 
> I don't think the approach in this PR is quite right. Furthermore, wrt #14373, I found this comment which provides reason to not add support for specifying subsystem types: [#6530 (comment)](https://github.com/canonical/lxd/pull/6530#issuecomment-561466424).
> 
> We'll have to rethink our solution here.

Which part of the comment makes you think multiple devices shouldn't be allowed?

**Comment by kadinsayani** [comment]
> Which part of the comment makes you think multiple devices shouldn't be allowed?

> If we don't specify a subsystem we end up enumerating multiple devices in loadUnixDevice(). For example, if we plug in a usb with a FS on it, we get a device for the usb itself and then a device for the sda device node entry as well since they share the same product/vendor IDs

**Comment by kadinsayani** [comment]
Looks like this was an issue raised in the original PR but was never actually addressed properly:

> The partitions (sda1, sda2 and sda3) only showed up during USB hotplug, on container start or when adding the device, only sda shows up

Also, this [commit](https://github.com/canonical/lxd/commit/a10be101fa4d9a0aedc9a24116d16df32c6f4457) added logic to not add a device with a minor number of 0, but for most usb devices, the first sub device has a minor number of 0.

**Comment by kadinsayani** [comment]
Thanks to @jonathan-conder for testing this locally with his camera devices. The camera devices showed up as expected, but a mouse with a matching vendor id also appeared. This is because when `udev_enumerate_add_match_property` is called more than once, each matching result is ORed - only at least one property needs to match.

It's now clear why this [commit](https://github.com/canonical/lxd/commit/a10be101fa4d9a0aedc9a24116d16df32c6f4457) added logic to not add a device with a major or minor number of 0. However, this approach doesn't allow all devices with a matching product or vendor id to be captured. For example, if we ignore devices with a minor number of 0, we may see cases like this:

```
$ ls -l /dev/video*
crw-rw----+ 1 root video 81, 0 Oct 19 05:29 /dev/video0
crw-rw----+ 1 root video 81, 1 Oct 19 05:29 /dev/video1
crw-rw----+ 1 root video 81, 2 Oct 19 05:41 /dev/video2
crw-rw----+ 1 root video 81, 3 Oct 19 05:41 /dev/video3
$ lxc exec u1 -- sh -c 'ls -l /dev/video*'
crw-rw---- 1 root root  81, 1 Oct 18 16:39 /dev/video1
crw-rw---- 1 root video 81, 2 Oct 18 16:41 /dev/video2
crw-rw---- 1 root video 81, 3 Oct 18 16:41 /dev/video3
```

We should see `/dev/video[0-3]`.

I think the best approach is to require specifying subsystem type along with either vendor or product id. Requiring subsystem type to be specified will lead to less "edge cases". @tomponline, what do you think?

**Comment by kadinsayani** [comment]
Or, we just need a way to filter on both vendor id and product id. This would also be helpful for https://github.com/canonical/lxd/issues/14373 so we could filter on subsystem && vendorid. I'll look into this further.

**Comment by kadinsayani** [comment]
I believe I've got a fix working. Just waiting on a test with camera devices and then this PR will be good to go.

**Comment by kadinsayani** [comment]
The fix is now confirmed to be working. We've noticed another inconsistency where `/dev/bus/usb/...` devices are excluded by what's added by `unix-hotplug`, however, the device(s) appear after replugging the device. [This comment](https://github.com/canonical/lxd/pull/6530#issuecomment-561466424) provides some insight into the original decision for excluding `/dev/bus/usb/...` devices.

**Comment by jonathan-conder** [comment]
As Kadin says, I'm happy with this version. For our use case it doesn't matter whether the USB device is added or not, but based on that comment I agree it makes sense to avoid adding the USB device when replugging

**Comment by kadinsayani** [comment]
I've added a commit which resolves the inconsistent device matching bug. `/dev/bus/usb/...` will be ignored when `unix-hotplug` devices are registered and started (replugged or after container restart). 

@tomponline this PR now fully resolves the issue.

File: lxd/device/unix_hotplug.go [hunk 3]
```diff
@@ -149,29 +148,36 @@ func (d *unixHotplug) Start() (*deviceConfig.RunConfig, error) {
...
 
-	device := d.loadUnixDevice()
-	if d.isRequired() && device == nil {
+	devices := d.loadUnixDevices()
+	if d.isRequired() && devices == nil {
 		return nil, fmt.Errorf("Required Unix Hotplug device not found")
 	}
 
-	if device == nil {
+	if devices == nil {
```
Comment (tomponline) [inline lxd/device/unix_hotplug.go:165]: do we need this anymore? If the devices var is nil or zero length then the for loop wont iterate it and the return statement at the bottom will run as usual.

File: lxd/device/unix_hotplug.go [hunk 4]
```diff
@@ -149,29 +148,36 @@ func (d *unixHotplug) Start() (*deviceConfig.RunConfig, error) {
 	runConf := deviceConfig.RunConfig{}
 	runConf.PostHooks = []func() error{d.Register}
 
-	device := d.loadUnixDevice()
-	if d.isRequired() && device == nil {
+	devices := d.loadUnixDevices()
+	if d.isRequired() && devices == nil {
```
Comment (tomponline) [inline lxd/device/unix_hotplug.go:161]: Can we do `if d.isRequired() && len(devices) <= 0 {` to catch the scenario where devices is non-nil but zero length?

File: lxd/device/unix_hotplug.go [hunk 5]
```diff
@@ -149,29 +148,36 @@ func (d *unixHotplug) Start() (*deviceConfig.RunConfig, error) {
...
 
-	// setup device
-	var err error
-	if device.Subsystem() == "block" {
-		err = unixDeviceSetupBlockNum(d.state, d.inst.DevicesPath(), "unix", d.name, d.config, major, minor, device.Devnode(), false, &runConf)
-	} else {
-		err = unixDeviceSetupCharNum(d.state, d.inst.DevicesPath(), "unix", d.name, d.config, major, minor, device.Devnode(), false, &runConf)
-	}
+		if err != nil {
+			err := unixDeviceRemove(d.inst.DevicesPath(), "unix", d.name, "", &runConf)
```
Comment (tomponline) [inline lxd/device/unix_hotplug.go:183]: Please can you use the `revert` package here.

This makes it easier to maintain in the future because we wont have to worry about reverting every step manually on any new error cases.

File: lxd/device/unix_hotplug.go [hunk 6]
```diff
@@ -203,9 +209,9 @@ func (d *unixHotplug) postStop() error {
 	return nil
 }
 
-// loadUnixDevice scans the host machine for unix devices with matching product/vendor ids
-// and returns the first matching device with the subsystem type char or block.
-func (d *unixHotplug) loadUnixDevice() *udev.Device {
+// loadUnixDevices scans the host machine for unix devices with matching product/vendor ids
+// and returns the matching devices with subsystem types of char or block.
+func (d *unixHotplug) loadUnixDevices() []*udev.Device {
```
Comment (tomponline) [inline lxd/device/unix_hotplug.go:223]: can we return just `[]udev.Device` ?

File: lxd/device/unix_hotplug.go [hunk 7]
```diff
@@ -230,27 +236,32 @@ func (d *unixHotplug) loadUnixDevice() *udev.Device {
...
 
 	devices, _ := e.Devices()
-	var device *udev.Device
+	var matchingDevices []*udev.Device
 	for i := range devices {
-		device = devices[i]
+		device := devices[i]
 
-		if device == nil {
+		if device == nil || device.Devnum().Major() == 0 || device.Devnode() == "" || strings.HasPrefix(device.Subsystem(), "usb") {
```
Comment (tomponline) [inline lxd/device/unix_hotplug.go]: Please can you add a comment here explaining the *why* for skipping in these cases.

File: lxd/device/unix_hotplug.go [hunk 8]
```diff
@@ -230,27 +236,32 @@ func (d *unixHotplug) loadUnixDevice() *udev.Device {
...
 			continue
 		}
 
-		devnum := device.Devnum()
-		if devnum.Major() == 0 || devnum.Minor() == 0 {
-			continue
+		match := false
+		vendorIDMatch := device.PropertyValue("ID_VENDOR_ID") == d.config["vendorid"]
```
Comment (tomponline) [inline lxd/device/unix_hotplug.go]: why is this needed when there is apparently filtering being setup via `e.AddMatchProperty` above?

**Review by tomponline — CHANGES_REQUESTED** [review]
A few minor tweaks and a question about filtering. Thanks

File: lxd/device/unix_hotplug.go (hunk 8 above)
Comment (kadinsayani) [reply lxd/device/unix_hotplug.go]: `e.AddMatchProperty` calls `udev_enumerate_add_match_property` under the hood. Each matching result is ORed - only at least one property needs to match.

File: lxd/device/unix_hotplug.go (hunk 3 above)
Comment (kadinsayani) [reply lxd/device/unix_hotplug.go:165]: You're right, we don't need this check anymore. Good catch.

File: lxd/device/unix_hotplug.go (hunk 4 above)
Comment (kadinsayani) [reply lxd/device/unix_hotplug.go:161]: Sure thing 👍

File: lxd/device/unix_hotplug.go (hunk 5 above)
Comment (kadinsayani) [reply lxd/device/unix_hotplug.go:183]: Sounds good

File: lxd/device/unix_hotplug.go (hunk 6 above)
Comment (kadinsayani) [reply lxd/device/unix_hotplug.go:223]: Yes, we can just just dereference `device` when we append it to `matchingDevices`.

File: lxd/device/unix_hotplug.go (hunk 7 above)
Comment (kadinsayani) [reply lxd/device/unix_hotplug.go]: Sure thing!

File: lxd/device/unix_hotplug.go (hunk 8 above)
Comment (kadinsayani) [reply lxd/device/unix_hotplug.go]: From the udevadm man page:

```
-p, --property-match=PROPERTY=VALUE
           Trigger events for devices with a matching property value. This option supports shell style pattern matching. When this option is specified more than once, then each matching result is
           ORed, that is, devices which have one of the specified properties are triggered.
```

I've removed the redundant `e.AddMatchProperty` calls since we have the new consolidated matching logic now.

Comment (tomponline) [reply lxd/device/unix_hotplug.go]: Ok i see. The AddMatchProperty calls should remain though to provide early push down filtering.

Comment (kadinsayani) [reply lxd/device/unix_hotplug.go]: Sounds good - I've added them back.

File: lxd/device/unix_hotplug.go [hunk 9]
```diff
@@ -149,31 +149,42 @@ func (d *unixHotplug) Start() (*deviceConfig.RunConfig, error) {
...
-	if device.Subsystem() == "block" {
-		err = unixDeviceSetupBlockNum(d.state, d.inst.DevicesPath(), "unix", d.name, d.config, major, minor, device.Devnode(), false, &runConf)
-	} else {
-		err = unixDeviceSetupCharNum(d.state, d.inst.DevicesPath(), "unix", d.name, d.config, major, minor, device.Devnode(), false, &runConf)
-	}
+		// Remove unix device on failure to setup device.
+		revert.Add(func() {
+			err = unixDeviceRemove(d.inst.DevicesPath(), "unix", d.name, "", &runConf)
+			if err != nil {
+				d.logger.Error("Failed to remove unix device", logger.Ctx{"device": d.name, "err": err})
```
Comment (tomponline) [inline lxd/device/unix_hotplug.go:188]: Perhaps "Failed to revert..." or "Failed to clean up..." would be clearer here, as otherwise users may think the LXD instance device itself was being removed even though it technically never got started. Also "removed" indicates the device was trying to be "added", but that is not whats happening from a user perspective, the device is actually being "started".

File: lxd/device/unix_hotplug.go [hunk 10]
```diff
@@ -149,31 +149,42 @@ func (d *unixHotplug) Start() (*deviceConfig.RunConfig, error) {
 	runConf := deviceConfig.RunConfig{}
 	runConf.PostHooks = []func() error{d.Register}
 
-	device := d.loadUnixDevice()
-	if d.isRequired() && device == nil {
-		return nil, fmt.Errorf("Required Unix Hotplug device not found")
+	devices := d.loadUnixDevices()
+	if d.isRequired() && len(devices) <= 0 {
+		return nil, fmt.Errorf("Required unix hotplug device %q not found", d.name)
```
[the remark this answers is not in the document]
Comment (tomponline) [reply lxd/device/unix_hotplug.go:164]: Ah wait, GH was munging the two changes together, this is in Start().

Still, I don't think we should be including the device name in the error.

Because we can see from the call site, returns like this:

```
return nil, fmt.Errorf("Failed to start device %q: %w", dev.Name(), err)
```

File: lxd/device/unix_hotplug.go [hunk 11]
```diff
@@ -149,31 +149,42 @@ func (d *unixHotplug) Start() (*deviceConfig.RunConfig, error) {
...
+			err = unixDeviceRemove(d.inst.DevicesPath(), "unix", d.name, "", &runConf)
+			if err != nil {
+				d.logger.Error("Failed to remove unix device", logger.Ctx{"device": d.name, "err": err})
+			}
+		})
 
-	if err != nil {
-		return nil, err
+		if err != nil {
+			return nil, fmt.Errorf("Unable to setup unix hotplug device %q: %w", d.name, err)
```
Comment (tomponline) [inline lxd/device/unix_hotplug.go:193]: No need for including device name in error, see above.

File: lxd/device/unix_hotplug.go [hunk 12]
```diff
@@ -230,27 +241,33 @@ func (d *unixHotplug) loadUnixDevice() *udev.Device {
...
 	devices, _ := e.Devices()
-	var device *udev.Device
+	var matchingDevices []udev.Device
 	for i := range devices {
-		device = devices[i]
+		device := devices[i]
 
-		if device == nil {
+		// Ignore USB bus devices (handled by `usb` device type) and devices without a valid devpath or major number.
+		if device == nil || device.Devnum().Major() == 0 || device.Devnode() == "" || strings.HasPrefix(device.Subsystem(), "usb") {
```
Comment (tomponline) [inline lxd/device/unix_hotplug.go]: youve added a comment for the *what* but not yet the *why* of the skipping logic here.

Comment (tomponline) [reply lxd/device/unix_hotplug.go]: the aim here is to document the git archeology you've spent a lot of time on to save future readers the work.

File: lxd/device/unix_hotplug.go (hunk 9 above)
Comment (kadinsayani) [reply lxd/device/unix_hotplug.go]: Agreed.

**Review by tomponline — APPROVED** [review]
